types/sanctuary/index.d.ts

```diff
@@ -107,19 +103,49 @@ declare namespace Sanctuary {
     gte<A>(x: Ord<A>): (y: Ord<A>) => boolean;
     min<A>(x: Ord<A>): (y: Ord<A>) => A;
     max<A>(x: Ord<A>): (y: Ord<A>) => A;
+    // This is ugly, but makes its job, 
+    // I originally tried to use union:
+    // clamp<A extends (Ord<A> | boolean | Date | number | string)>(a: A): (b: A) => (c: A) => A;
+    // in which case we would need only three overloads (for simple, array and StrMap)
+    // but then the following case fails:
+    // clamp(2)(3)
+    // with error "'3' is not assignable to 2". Because of type inference type 
+    // parameter A was narrowed down to type '2' (type of literal '2')
+    // and not to type 'number'. This would force client to provide explicit
+    // type parameter each time (to prevent narrowing), e.g. clamp<number>(2)(3)
+    clamp(a: boolean): (b: boolean) => (c: boolean) => boolean;
+    clamp(a: boolean[]): (b: boolean[]) => (c: boolean[]) => boolean[];
+    clamp(a: StrMap<boolean>): (b: StrMap<boolean>) => (c: StrMap<boolean>) => StrMap<boolean>;
+    clamp(a: Date): (b: Date) => (c: Date) => Date;
+    clamp(a: Date[]): (b: Date[]) => (c: Date[]) => Date[];
+    clamp(a: StrMap<Date>): (b: StrMap<Date>) => (c: StrMap<Date>) => StrMap<Date>;
+    clamp(a: number): (b: number) => (c: number) => number;
+    clamp(a: number[]): (b: number[]) => (c: number[]) => number[];
```

Comment:
What about `number[][]`, `number[][][]`, et cetera?

Reply:
That's a good question

Reply:
I don't quite understand - why does `clamp` work on arrays? I mean, does `Array` conform to fantasy land `Ord` specification? If so, we could simply declare it _somehow_ (interfaces merging? Not sure, but there is probably a way) and then it will be solved automagically by overload in line 128.

Reply:
> I don't quite understand - why does `clamp` work on arrays? I mean, does `Array` conform to fantasy land `Ord` specification?

Yes, it does. The semantics match those of Haskell's `Ord a => Ord [a]` instance:

```haskell
Prelude> ["foo"] <= ["foo"]
True

Prelude> ["foo"] <= ["bar"]
False

Prelude> ["foo"] <= ["foo", "bar"]
True
```

I hope we're not faced with a choice between writing hundreds of definitions (and still rejecting some valid programs) and resorting to using the empty interface. :crossed_fingers:

Reply:
I am sure we can do better than that. As I said, merging interfaces could be one answer. I got it to work on a deliberately [simplified example](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgMIHsC2AHOU4BGANhADwAqAfMgN4BQyyCWuUE56AFOmABbQAuZOQCUQgIzIAPsgAM05AFpxdAL506Aek3IA7igDkAEwgIieCAeQBBKPgCeyMOmTAcJTBHBoWeQiVdwaHgkABo9YD50AFcwZGwodAA3YCNQAHNXdwhPcDgwYHQQOlBIWEQUWwcKaggAD0gQIwBnHxw-YjIquHsa6hp1OmYQZrjmdrZkAF5kCmR6xpa21n8yKkpOOqFRaepOe22RXeQ6gDpx1nYuexENbWFeYFailFxm5ohWuFb67FNIIxDXxsTgAbXE4QATABdERggDM4QALLC7jo4E0nI9niAUPBgEQvj86n8EACgRMIGCIcgYXDQYjkKCUaitDp9EwMchcWTkN9kJgMY48A4iXoULoCUQKZcweCodDYXL4dDwszFSIgA), but couldn't get it to work with Sanctuary (I am not really sure how does merging work with exported types, it's probably just a matter of putting this "declaration" in a correct place).

Reply:
Updating `interface Setoid<A> {}` would be a good starting point, I think.

Reply:
I think I came up with a [solution](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgMIHsC2AHOU4BGANhADwAqAfMgN4BQyyCWuUE56AFOmABbQAuZOQCUQkAFdMBaHQC+dOmACe2FAHkoAE2QBeZAXToScEMgA+yACJxIF5JOnR7AZzBRQAc3sYceQiSkpsrUlppa0BBaAIJQ+Mp0oJCwiBrakTFxcMrIEAAekCBaLsix8aTh1DQKigD0tcjoAG7QROhwxcgw6FC5LVA52B6YwGDALcgqanQwEiAIY+hmzH5snHniUjJQYsicyptOO3rUjtsA3DNzC8BLTCx4EPtCbh4gnrvPyK9eIicOW2gl1m80Wyweaw2BiMJhAnwO0OMEFMf10p0BUGB1zB91WTyhNkg8KEhIgqPRR0udHqyFMOk8EBA0GACEa-TaHSuoNu4LxFFyBUZnUq6yEoiEX1E-zOQK5NzuK1YZHIAsKwu0lFFwmJ2ulGNoDGQbDAEigZhB8pAkr+9EYjGNprMAAZLowFHIqTTyLxgCUlihcC4XBASnASvk1AsonRFY9OABtACMABpkAAmAC6IgTAGZUwAWLN1Bp0yY+v1MrpwYBEUPhvKRyBaGMQp5J1OZ7PxvPIeOFovUhoAdxQCFMDhDYFpJUwwVpWWUdeQI+XNaILbxCfb6YzWa3OYzqb7u5EihjRDDJQAYkYDYxY2wONw+IJkDf0DbDfaICazcgXYaNSMN06AAMqRsAMAspwn52kaP6Osg+Zpq6yDuooCAXkGyAAEJ4HeuJKk+PD8FAQh4cctp2g6f4AW68jFlWNZLhGEBRloqYuKASBvrepYUfOKAgDwhG2MABA1ignBaBIKBgOgXRGOBbGQSyyCYD+vDoFop4Pk8TJDrxXAiNmBm4XgMGnoOyCBsGLgbkqCZme+MF7vGzlGK5Vk0rZIYOXGibZjm3kNPAzHTgKjbRnpnCBZwADkhAIPFVlnkAA) (and it applies not only to this specific case, it might work in general, I think), at least I can't think of an example that would break it, do you?

Reply:
I did my best to implement your suggestion for Setoid and Ord. I have pushed a commit. In addition to being irked by having to write `S.equals<string> ('abc') ('def')`, I encountered one problem. The following expression is invalid, but `tsc` does not report an error:

```javascript
S.equals (S.Just ('abc')) (S.Just (123))
```

Could we prohibit the above without adding a special case for `Maybe<A>`?